controller/network/network.go

```diff
@@ -243,100 +208,84 @@ func (network *Network) LinkChanged(l *Link) {
 	}()
 }
 
-func (network *Network) BindService(srcR *Router, token string, serviceId string, peerData map[uint32][]byte) error {
+func (network *Network) CreateSession(srcR *Router, clientId *identity.TokenId, serviceId string) (*session, error) {
 	log := pfxlog.Logger()
 
 	// 1: Find Service
-	if svc, found := network.serviceController.get(serviceId); found {
-		svc.Egress = srcR.Id
-		svc.EndpointAddress = "hosted:" + token
-		svc.PeerData = peerData
-
-		log.Debugf("binding service[%s] to session[%s] with hostdata[%v]", serviceId, token, svc.PeerData)
-		return network.serviceController.update(svc)
+	svc, err := network.Services.Read(serviceId)
+	if err != nil {
+		return nil, err
 	}
-	return errors.New("invalid service")
-}
-
-func (network *Network) UnbindService(srcR *Router, token string, serviceId string) error {
-	// 1: Find Service
-	if svc, found := network.serviceController.get(serviceId); found {
-		svc.Egress = network.nodeId.Token // TODO: can't set this to null, so setting it to the controller ID
-		svc.EndpointAddress = "hosted:unbound"
-		svc.PeerData = nil
 
-		return network.serviceController.update(svc)
+	// 2: Allocate Session Identifier
+	sessionIdHash, err := network.sequence.NextHash()
+	if err != nil {
+		return nil, err
 	}
-	return errors.New("invalid service")
-}
+	sessionId := &identity.TokenId{Token: sessionIdHash}
 
-func (network *Network) CreateSession(srcR *Router, clientId *identity.TokenId, serviceId string) (*session, error) {
-	log := pfxlog.Logger()
+	if len(svc.Endpoints) == 0 {
+		return nil, errors2.Errorf("service %v has no Endpoints", serviceId)
+	}
 
-	// 1: Find Service
-	if svc, found := network.serviceController.get(serviceId); found {
-		// 2: Allocate Session Identifier
-		sessionIdHash, err := network.sequence.NextHash()
-		if err != nil {
-			return nil, err
-		}
-		sessionId := &identity.TokenId{Token: sessionIdHash}
+	// 3: select endpoint
+	endpoint := svc.Endpoints[0]
```

Comment:
Selecting an endpoint before a router is selected _will_ have impacts on smart routing in general. We could very easily implement a strategy that always ends up selecting a poorly-performing router.

Not sure if there's anything to do right this second, other than just be aware that we're potentially neutering smart routing in its current form. The router selection process is being preempted by endpoint selection.

Reply:
An alternative may be something along the lines of: smart routing looks at the endpoints for a service, and then ranks those in performance order. The strategy would take that list as an input, and could override it, or it could choose to allow it to suggest a choice.

Reply:
I think we've got several options, including passing implicit or explicit weights from smart routing, as well as allowing endpoint selection to be deferred to smart routing, possibly by flipping the paradigm and having endpoint selection return a set of acceptable endpoints and then letting smart routing take over.

Reply:
Indeed. I'm sure we can figure out something that will work.